### PR TITLE
Add script type and script name to error messages

### LINKED ISSUE #6653: Script engines should be aware of the script name
It would be nice if we could access the name of the script from the script engine framework so it could be logged when errors occur. This would be especially helpful for additional logging for users with dynamic scripting disabled.

### PR DESCRIPTION
**Pull request by jdconrad**
Modified ScriptEngineService to pass in a CompiledScript object
to several methods with newly added name and type member variables.
This can in turn be used to give better scripting error messages
with the type of script used and the name of the script.

Required slight modifications to the caching mechanism.

Note that this does not enforce good behavior in that plugins will
have to write exceptions that also output the name of the script
in order to be effective.  There was no way to wrap the script
methods in a try/catch block properly further up the chain because
many have script-like objects passed back that can be run at a
later time.

fixes #6653

### CODE REVIEW

File: src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java [hunk 1]
```diff
@@ -98,22 +99,26 @@ public Object compile(String template) {
...
+                try {
+                    writer.close();
+                } catch (IOException e) {
+                    logger.error("Could not execute query template [" + template.name() + "] (failed to close writer): ", e);
+                }
             }
+            return result.bytes();
+        } catch (Exception e) {
+            throw new ScriptException("Error with query template [" + template.name() + "]: " + e.getMessage(), e);
         }
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java]: Can we simplify this to use try-with-resources? close() should call flush() anyway, and then this way no exceptions are ever hidden, instead they are chained with suppression: e.g.

    try (UTF8StreamWriter writer = ...) {
      execute();
    } catch (Exception e) {
      logger.error(xxx);
      throw new ScriptException(....)
    }

File: src/main/java/org/elasticsearch/script/expression/ExpressionScript.java [hunk 2]
```diff
@@ -109,7 +115,7 @@ public void setNextVar(String name, Object value) {
                 if (value instanceof Number) {
                     specialValue.setValue(((Number)value).doubleValue());
                 } else {
-                    throw new ExpressionScriptExecutionException("Cannot use expression with text variable");
+                    throw new ExpressionScriptExecutionException("Error with " + compiledScript.type().name() + " script [" + compiledScript.name() + "]: Cannot use expression with text variable");
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/script/expression/ExpressionScript.java]: @jdconrad would it make sense to have a `toString` methdo on CompiledScript that includes name and type?

File: src/main/java/org/elasticsearch/script/CompiledScript.java [hunk 3]
```diff
@@ -24,20 +24,42 @@
...
 
     /**
      * Constructor for CompiledScript.
+     * @param type The type of script to be executed.
+     * @param name The name of the script to be executed.
      * @param lang The language of the script to be executed.
      * @param compiled The compiled script Object that is executable.
      */
-    public CompiledScript(String lang, Object compiled) {
+    public CompiledScript(ScriptService.ScriptType type, String name, String lang, Object compiled) {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/CompiledScript.java]: Would it make sense to move ScriptType out of ScriptService at some point (doesn't need to be in this PR necessarily)

File: src/main/java/org/elasticsearch/script/ScriptService.java [hunk 4]
```diff
@@ -263,10 +263,12 @@ public CompiledScript compileInternal(Script script) {
             return compiled;
         }
 
+        String name = "";
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: can the name end up being empty?

File: src/main/java/org/elasticsearch/script/ScriptService.java [hunk 5]
```diff
@@ -274,11 +276,26 @@ public CompiledScript compileInternal(Script script) {
         CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
             //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+            try {
+                compiled = new CompiledScript(script.getType(), name, lang, scriptEngineService.compile(code));
+            } catch (RuntimeException re) {
+                throw new RuntimeException("Compilation error with " + script.getType().name()
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: ScriptType has a nice `toString`, not sure you should call `name()` to print it out

File: src/main/java/org/elasticsearch/script/ScriptService.java [hunk 6]
```diff
@@ -274,11 +276,26 @@ public CompiledScript compileInternal(Script script) {
...
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
+        if (compiled.type() != script.getType() || name != null && !name.equals(compiled.name())) {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: can the name ever be null here?

File: src/main/java/org/elasticsearch/script/CompiledScript.java (hunk 3 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/CompiledScript.java]: I think it probably does.  I will opt to leave this for another PR.  We may want to look at lot of the smaller classes in ScriptService as well.  Indexed script for instance is only used in a single place as far as I can tell, so maybe we convert that to a method to get the indexed script name out of the url?

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 4 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: For this and the comment below about null -- this was two different lines of thinking, and I forgot to change the latter.  I was thinking for inline scripts I would leave the string empty.  Given Simon's suggestion of a toString method, I'm going to go back to null.

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 6 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: See the comment above about this issue.

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 5 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: Oops.  I missed that.  This is a better way to do this, and I will make the change.

File: src/main/java/org/elasticsearch/script/expression/ExpressionScript.java (hunk 2 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/expression/ExpressionScript.java]: Yes, thanks for the suggestion.  I will make the change.

File: src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java (hunk 1 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java]: Changed.

**Comment by jdconrad** [comment]
Thanks to @rmuir @s1monw @javanna for the reviews -- and I'm ready for the next round.  I pushed a new commit with changes based on the prior comments, and I tried to add a bit more consistency to the modified error messages.

File: src/main/java/org/elasticsearch/script/ScriptService.java [hunk 7]
```diff
@@ -263,10 +264,13 @@ public CompiledScript compileInternal(Script script) {
             return compiled;
         }
 
+        // Name will remain null for inline scripts.
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: is this comment correct? seems to me that we assign it only in case of indexed scripts?

File: src/main/java/org/elasticsearch/script/ScriptService.java [hunk 8]
```diff
@@ -274,11 +278,30 @@ public CompiledScript compileInternal(Script script) {
...
+            }
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
```
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: I find this new code block confusing, can we maybe add the script type to the cache key instead or something along those lines?

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 7 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: I skipped file scripts here, as if it was a file script it had already been processed, but added a note to the comment to clarify.

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 8 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: @javanna I see three options here:

1. Decide to leave it as is with the benefit being this is the least invasive; however, it is complicated.

2. Add name to the cache key (instead of script type) because there are two cases currently where we can confuse different scripts described in the comment.  The first is if the code is the same between an inline script and an indexed script, you may get back a compiled script with incorrect name and type information.  The second is if two indexed scripts have the same code but different names, you may get back a compiled script with an incorrect name.  Adding name to the cache key would solve both problems.

3. We could leave the cache key as is and cache only the executable script objects instead of compiled script objects.  Then we create a new compiled script object to be returned with the current metadata (name, type, and lang).  I see the point of the cache as trying to prevent compiling the same script over and over, so this solution solves that problem and might actually be the cleanest way of doing this.

Let me know what you think, and then I'll implement the solution we decide upon.  Thanks!

**Comment by jdconrad** [comment]
@javanna Please take a look at the response I left you when you get a chance.

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 7 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: gotcha

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 8 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: I guess the question is whether it's a problem that two scripts with same code end up being one single entry in the cache, beyond side effect of returning wrong name etc. (which is a bug). If so we should go for 2) and add all the needed info to the cache key so we can distinguish those "same" scripts. Otherwise we could go with 3) too. Both 2. and 3. sound ok to me, which one do you prefer?

**Comment by javanna** [comment]
thanks @jdconrad I replied, this looks good on my end besides the comments I left, maybe @uboness can have a look too?

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 6 above)
Comment (rmuir) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: For a boolean expression this complicated i would turn negative checks into positive ones:

    if (compiled.type() == script.getType() && (name == null || name.equals(compiled.name())) {

I would also try to use locals to break it down easier. Something like:

    boolean sameType = 
    boolean sameName =

    if (sameName && sameType) {

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 8 above)
Comment (rjernst) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: I don't quite follow all the options here, but why is it not as simple as type + "value"?
`file_filename`
`indexed_name`
`inline_a+b`

Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: @rjernst I'll give this some thought.  Maybe it makes sense to turn cache key into a method that just returns a concatenated string instead of a full blown object that isn't really necessary?

File: src/main/java/org/elasticsearch/script/ScriptService.java (hunk 6 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/script/ScriptService.java]: @rmuir Thanks for the feedback -- I don't think this code is going to remain once the cache key is changed, but if it does I will make sure it's simplified in this form.

**Comment by javanna** [comment]
what's the status on this? @uboness did you have time to have a look?

**Comment by jdconrad** [comment]
@javanna No update right now.  I have been caught up in other issues and need to come back to this.  I like @rjernst 's suggestion of making the CacheKey class into a simple string with all the necessary information to make the key unique as that's what's ultimately happening anyway.  What do you think?

**Comment by javanna** [comment]
I think it would be nice if we could get this in, especially given that it's marked breaking. I wouldn't want this PR to be forgotten for 2.0.

**Comment by jdconrad** [comment]
@javanna I will try to get it done before the end of the week then.

**Comment by jdconrad** [comment]
@rjernst Oops, meant this PR :)

File: core/src/main/java/org/elasticsearch/script/CompiledScript.java [hunk 9]
```diff
@@ -24,17 +24,39 @@
...
 public class CompiledScript {
 
+    private final ScriptService.ScriptType type;
+    private final String name;
     private final String lang;
     private final Object compiled;
 
     /**
      * Constructor for CompiledScript.
+     +     +     * @param type The type of script to be executed.
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/CompiledScript.java]: Looks like a merge issue? (the '+'s)

File: core/src/main/java/org/elasticsearch/script/CompiledScript.java [hunk 10]
```diff
@@ -24,17 +24,39 @@
...
+     +     +     * @param name The name of the script to be executed.
      * @param lang The language of the script to be executed.
      * @param compiled The compiled script Object that is executable.
      */
-    public CompiledScript(String lang, Object compiled) {
-        this.lang = lang;
-        this.compiled = compiled;
+    public CompiledScript(ScriptService.ScriptType type, String name, String lang, Object compiled) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/CompiledScript.java]: can we move ScriptType out from ScriptService?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 11]
```diff
@@ -249,50 +249,70 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
...
         if (logger.isTraceEnabled()) {
-            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, script.getType(), script.getScript());
+            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, type, name);
         }
 
         ScriptEngineService scriptEngineService = getScriptEngineServiceForLang(lang);
-        String cacheKey = getCacheKey(scriptEngineService, script.getScript());
+        String cacheKey;
+
+        CompiledScript compiledScript;
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: I would move this into the FILE block, and redefine later when it is needed, so the FILE block is clearly on its own? same with cacheKey?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 12]
```diff
@@ -249,50 +249,70 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
...
+
+            return compiledScript;
         }
 
+        //script.getScript() will be code if the script type is inline
         String code = script.getScript();
 
         if (script.getType() == ScriptType.INDEXED) {
-            final IndexedScript indexedScript = new IndexedScript(lang, script.getScript());
+            final IndexedScript indexedScript = new IndexedScript(lang, name);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: maybe add a comment here that we must lookup the indexed script before the cache because the code may have changed?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 13]
```diff
@@ -249,50 +249,70 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
...
+        //if the script type is inline the name will be the same as the code for identification in exceptions
+        cacheKey = getCacheKey(scriptEngineService, name, code);
+        compiledScript = cache.getIfPresent(cacheKey);
+
+        if (compiledScript == null) {
+            //Either an un-cached inline script or indexed script
+            try {
+                compiledScript = new CompiledScript(type, name, lang, scriptEngineService.compile(code));
+            } catch (Exception exception) {
+                compiledScript = new CompiledScript(type, name, lang, null);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: Can we just print out the pieces type/name/lang here without creating a compiled script just to do a toString()?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 14]
```diff
@@ -598,9 +618,9 @@ public String toString() {
         }
     }
 
-    private static String getCacheKey(ScriptEngineService scriptEngineService, String script) {
+    private static String getCacheKey(ScriptEngineService scriptEngineService, String name, String code) {
         String lang = scriptEngineService.types()[0];
-        return lang + ":" + script;
+        return lang + ":" + name + (code != null ? ":" + code : "");
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: In the case of an inline script, won't `name` already be the entire script? So in this case the cache key would be the code duplicated twice?

**Comment by rjernst** [comment]
LGTM, I left a couple small suggestions.

File: core/src/main/java/org/elasticsearch/script/CompiledScript.java (hunk 9 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/CompiledScript.java]: Oops.  Merge error from using a patch between the pre and post plugin era.  Fixed.

File: core/src/main/java/org/elasticsearch/script/CompiledScript.java (hunk 10 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/CompiledScript.java]: Will do so in a different PR as discussed previously in this PR.

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 11 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: Done.

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 12 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: Done.

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 13 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: Done.

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 14 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: While this shouldn't affect uniqueness, to avoid confusion I will make it so inline scripts end up being lang:code.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review.
